platform/darwin/src/NSExpression+MGLAdditions.mm

```diff
@@ -659,11 +668,24 @@ + (instancetype)expressionWithMGLJSONObject:(id)object {
     if ([object isKindOfClass:[NSString class]] ||
         [object isKindOfClass:[NSNumber class]] ||
         [object isKindOfClass:[NSValue class]] ||
-        [object isKindOfClass:[MGLColor class]]) {
+        [object isKindOfClass:[MGLColor class]] ||
+        [object isKindOfClass:[MGLShape class]]) {
         return [NSExpression expressionForConstantValue:object];
     }
     
     if ([object isKindOfClass:[NSDictionary class]]) {
+        if (object[@"type"]) {
+            NSError *error;
+            NSData *shapeData = [NSJSONSerialization dataWithJSONObject:object options:0 error:&error];
+            MGLShape *shape;
+            if (shapeData && !error) {
+                shape = [MGLShape shapeWithData:shapeData encoding:NSUTF8StringEncoding error:&error];
```

Comment:
This works, but there may be a more efficient way to convert this JSON-like NSDictionary structure into an MGLShape. Perhaps we could first convert the NSDictionary structure into a C++ representation of the GeoJSON object, then convert to an MGLShape.

Reply:
Is there any performance considerations for bridging into an C++ object then back to `MGLShape`? Is this something that is going to be called frequently?

Reply:
In the current implementation, `MGLStyleValueTransformer::PropertyExpressionEvaluator` converts the C++ `mapbox::geojson` struct to a GeoJSON-formatted NSDictionary, then this method converts it to a string, then back to a C++ `mapbox::geojson` struct, then finally to an MGLShape object. It is more roundabout than it should be.

This conversion process happens once when setting a layer’s predicate or paint/layout property (which can happen automatically e.g. when localizing the style). mbgl will evaluate the already-converted expression on each feature, but no SDK-side code will run as part of the rendering loop.

Ideally, mbgl would preserve the `mapbox::geojson` struct so that `MGLJSONObjectFromMBGLValue()` could convert it directly to an MGLShape. Absent such a change in mbgl, there are two alternative, purely SDK-side approaches:

* Convert directly to C++ `mapbox::geojson` structs then to MGLShape, skipping NSJSONSerialization
* Directly initialize MGLShapes using new initializers that parse the NSDictionary, skipping C++

but I’m inclined to treat either optimization as tail work for when we know more about how this expression operator will be used. The performance characteristics could differ depending on the complexity of the passed-in GeoJSON feature.